### ISSUE TITLE
bot-Tag may appear in pattern

### ISSUE THREAD

**Issue by Josef-Meier** [Program-O/Program-O#442]
Version 2.6.10
Hello,
I have discovered some AIML files (in the header version 1.0) in which the bot tag is used not only in the template- but also in the pattern elements.
Like this, because I am building a German-speaking bot:   http://alicebot.wikidot.com/aiml:de-de:cdrossman:alice:german-2-aiml
It was not easy for me to find something more detailed about the bot tag under the AIML 1.0 standard or a specification. So I'm not sure if this actually complies with the specification of AIML 1.0, as some say.
Then I found a 10 year old forum entry, where the topic was treated:
   http://knytetrypper.proboards.com/thread/397/aiml-bot-name-pattern
Here are some dead links referenced on the pages of alicebot.org.
Here the bot-day should be described closer.
Therefore, it should already be possible with the AIML 1.0 standard to use the ```<bot name = "name" />```
 in "pattern". And only this ```<bot name = "name" />``` tag may be used in pattern.
But then I found another link where the bot tags are explained:
  http://mctarek.free.fr/AIML/AIML_Tags.htm
Here is a text extract of it:
```
 AIML 1.0 Tags Set
  Adopted by A.L.I.C.E. A.I. Foundation 		
	AIML Architecture Committee 		
	July 16, 2001
	
  Tag:                  Tag Type:                      Note:
  <bot name="name"/>    Built-in bot parameter         may appear in pattern
  <bot name="XXX"/>     Custom bot parameter
```
Here I have an AIML example:
Note: The bot name is unimportant in this case and does not need to be written into "pattern" because the name is resolved via the bot tag.
Otherwise you would have to adapt the import of foreign AIML files always the bot name.
```
<category>
   <pattern>DO YOU KNOW WHO * IS</pattern>
   <template>I don't know who <star/> is.</template>
</category>
<category>
   <pattern>DO YOU KNOW WHO <bot name="name"/> IS</pattern>
   <template>This is me. I am <bot name="name"/>.</template>
</category>
```
A small change in the function find_aiml_matches () of the file "find_aiml.php" may be help:
```
...
function find_aiml_matches(&$convoArr)
{
    global $dbn, $error_response;
    $user_id = $convoArr['conversation']['user_id'];
    runDebug(__FILE__, __FUNCTION__, __LINE__, "Finding the aiml matches from the DB", 4);

    $i = 0;
    //TODO convert to get_it
    $bot_id = $convoArr['conversation']['bot_id'];
    $bot_parent_id = $convoArr['conversation']['bot_parent_id'];
    runDebug(__FILE__, __FUNCTION__, __LINE__, "Bot ID = $bot_id. Bot Parent ID = $bot_parent_id.", 4);
    // get bot and bot parent ID's

    $default_aiml_pattern = $convoArr['conversation']['default_aiml_pattern'];

    #$lookingfor = get_convo_var($convoArr,"aiml","lookingfor");
    $convoArr['aiml']['lookingfor'] = str_replace('  ', ' ', $convoArr['aiml']['lookingfor']);
    $lookingfor = trim(_strtoupper($convoArr['aiml']['lookingfor']));

    //get the stored topic
    $storedtopic = trim(_strtoupper(get_topic($convoArr)));
    runDebug(__FILE__, __FUNCTION__, __LINE__, "Stored topic = '$storedtopic'", 4);

    //get the cleaned previous bot response (that)
    $lastthat = (isset ($convoArr['that'][1][1])) ? $convoArr['that'][1][1] : '';
    $lastthat = rtrim(_strtoupper($lastthat));

    $params = array();
    if ($bot_parent_id != 0 && $bot_parent_id != $bot_id)
    {
        $sql_bot_like = "(bot_id = :bot_id OR bot_id = :bot_parent_id)";
        $params[':bot_id'] = $bot_id;
        $params[':bot_parent_id'] = $bot_parent_id;
    }
    else {
        $sql_bot_like = "bot_id = :bot_id";
        $params[':bot_id'] = $bot_id;
    }
    // Build the pattern search
    $rplTemplate = "'[search]' LIKE (REPLACE(REPLACE(`[field]`, '*', '%'), '_', '%'))";

    //// Meier: "<bot name="name"/>" may appear in pattern
    //   http://mctarek.free.fr/AIML/AIML_Tags.htm
    //   old code
    // $rplTemplate = "'[search]' LIKE (REPLACE(REPLACE(`[field]`, '*', '%'), '_', '%'))";
    // start new code
    $bot_name = $convoArr['bot_properties']['name'];
    $rplTemplateBotName = "'[search]' LIKE (REPLACE(REPLACE(REPLACE(`[field]`, '*', '%'), '_', '%'),'<BOT NAME=\"NAME\"/>', '${bot_name}'))";
    // end new code

    $pattern_like = "\n        " . str_replace('[search]', $lookingfor, $rplTemplateBotName);
    $pattern_like = str_replace('[field]', 'pattern', $pattern_like);

...
```
Unfortunately, the import of an AIML with bot tag in the pattern does not work.
Something else has to be done too. Maybe this is even more difficult. 
Sorry for my bad English language skills.

Thanks

**Comment by OverlanAI** [comment]
Well you are doing well at posting in English and I am having no problems understanding.
As to the change I need to think about that because at one point Program-O did do it that way but when it came to multi-bots it created a new problem.
If you are just running one bot and never plan to have more than one (bot-id) then sure why not.
I am at a loss to exactly what the problem was back then other than it has to do with having many bots.

As for using a botname in a pattern goes, you can do it but you have to cheat a bit.
First i set the bots name using a code like "MY-NAME"  so it would never show up as a valid word in the pattern.
When ever my script sees the code "MY-NAME"  it knows to always swap it out with the real name.
Because the code i give also propagates every part of any AIML bot-name that gives me an advantage.
Next why Pattern is filtered from most sources from adding in say a "-" it is not totally filtered by the admin aiml edit so i can use a code like "MY-NAME" as a pattern code that it will process and is also ignored by everything else. That gave me a constant as far as program-O cares.
I also found the same trick works in THAT pattern and TOPIC (not tested).
But for me that works fine it would come out well with out an external script swapping out the "MY-NAME code with the Intended name of say "BILL". so why we can cheat it will not work in all cases.
I felt that approach is safe to do because the only one that add something like that in is the bot master anyway.
Your solution for the same problem i would need to do a lot of tests on because of the risk aiml loops forming and there are a few other conditions where that can cause odd results.
Overall i would agree that it adds a quite useful and powerful dimension to what we already have.
But the little voice in my ear is yelling "Not so fast there are problems to this".
TNX

**Comment by OverlanAI** [comment]
Well the only thing i can see is (forgetting multi-bot for a bit) is that a risk of Mysql injection when a file from unknown is uploaded, but that is a guess right now.
And yes if we do that then export needs to also leave it alone but i doubt it filters it out in that direction anyway. (my trick of adding a - into the pattern after the fact is also in the export) so i dont see why it would filter it out from the export anyway.
But as i look a bit more at what you have i see we could also have
<bot name="XXX"/>  Custom bot parameter.
And to me thats more powerful because we can set that to be whatever we want as long as we have it set in the bots profile.
So now im thinking why not go the next step and make it <bot name="whatever bot param we want"/>?
Why limit it to just <bot name="name"/> only?
Anyway food for thought and i like the idea but what are our risks?
TNX

**Comment by OverlanAI** [comment]
Now what we could do is to add in a bot profile called Misc.
so we have the name (fixed) as you have it but also allow for misc that is alos (fixed).
Now what that gives us, is what you have and a new one than can be used for anything else we might like to include in a pattern search.
That may be quite useful too because the value of (misc) and be anything the bot master wants it be and since it is used as constant quite easy to change or ignore if it is null.
So ill stand on your shoulders a bit and include a Custom bot parameter as misc and see what tricks that lets me do :)
tnx (great stuff man, i love progress)

**Comment by OverlanAI** [comment]
`$bot_name = $convoArr['bot_properties']['name'];`

You will want to do a check to make sure that name is not NULL and that will happen if the profile value has not been set yet as when we make a new bot  But I would feel better if it did a check there anyway just in case. I suppose we could just give it a value like SPACE or something if it does find it null.
Also i did not look to see if it was already handled some place else.

**Comment by OverlanAI** [comment]
`$rplTemplateBotName =`   i assume you mean `$rplTemplate =` ?
If i use the first one it breaks.

**Comment by Josef-Meier** [comment]
@OverlanAI 
No should be different:
- $rplTemplateBotName = "'[search]' LIKE (REPLACE(REPLACE(REPLACE(`[field]`, '*', '%'), '_', '%'),'<BOT NAME=\"NAME\"/>', '${bot_name}'))";
- $rplTemplate = "'[search]' LIKE (REPLACE(REPLACE(`[field]`, '*', '%'), '_', '%'))";
The code seems correct but the comment line for the code end was in the wrong position.
So you probably did not change the line after the end marker
   $pattern_like = "\n        " . str_replace('[search]', $lookingfor, $rplTemplateBotName);
I changed it above.
Sorry it was my mistake. 
Thanks

**Comment by OverlanAI** [comment]
Well i got it in there and it adds it ok but in the end it does not work but there is another problem with this and that is it scoring it somehow.
If i have something like your example and also have a more common * value, The star value match overrides the reply anyway.
* has a value but bot name= has none.
A simple atomic score would also override it.
So to make this actually useful you need to some how include a score value for it.
So yes we can get away with it but there needs to be more to it or it wont use it due to no score.
TNX

**Comment by OverlanAI** [comment]
Cool ok it works but there is still going to be a score problem.
The easy way to test it is just
 
`<category>
   <pattern><bot name="name"/></pattern>
   <template>This is me. I am <bot name="name"/>.</template>
</category>`

That way we know that it cannot find anything else but what we are testing for now.
What it looks like its doing is only giving us a word and direct match score only.
So if thats correct then almost any other close match in the pattern search will score way higher and not use the cool function we have right now.
I have an idea but for now it is a good start and we can run other tests now in a basic way at least.
TNX

**Comment by Josef-Meier** [comment]
Hello
Have worked out a few test cases for this topic.
My exported test cases for the "<BOT NAME="NAME"/>-tag in pattern" topic :
```
<?xml version="1.0" encoding="UTF-8"?>
<!--  file: test_2.aiml                                                 -->
<aiml version="1.0">
  <category>
    <pattern>YOUR NAME IS *</pattern>
    <template>That's wrong. My name is <bot name="name"/>.</template>
  </category>
  <category>
    <pattern>YOUR NAME IS *</pattern>
    <that>* GUESS WHAT MY NAME IS</that>
    <template>Wrong guess. My name is <bot name="name"/>.</template>
  </category>
  <category>
    <pattern>
      <bot name="name"/>
    </pattern>
    <that>* GUESS WHAT MY NAME IS</that>
    <template>Yes. That's right.</template>
  </category>
  <category>
    <pattern>YOUR NAME IS <bot name="name"/></pattern>
    <that>* GUESS WHAT MY NAME IS</that>
    <template>That's right.</template>
  </category>
  <category>
    <pattern>HELLO MY NAME IS *</pattern>
    <template><think><set name="name"><star/></set></think>Nice to meet you, <get name="name"/>. Guess what my name is?</template>
  </category>
  <category>
    <pattern><bot name="name"/> IS A NICE NAME</pattern>
    <template>Nice, that you like the name <bot name="name"/>.</template>
  </category>
  <category>
    <pattern>DO YOU KNOW WHO JOE IS</pattern>
    <template>Joe is my neighbor.</template>
  </category>
  <category>
    <pattern>DO YOU KNOW WHO * IS</pattern>
    <template>I don't know who <star/> is.</template>
  </category>
  <category>
    <pattern>DO YOU KNOW WHO <bot name="name"/> IS</pattern>
    <template>This is me. I am <bot name="name"/>.</template>
  </category>
  <category>
    <pattern>
      <bot name="name"/>
    </pattern>
    <template>Yes. This is me. I am <bot name="name"/>.</template>
  </category>
</aiml>
```
Test Dialog:
  
Note: Bot name is "Chalmun". The canteen owner from Star Wars :-)
And I copied the dialogue and inserted it here.

User say: Hello, my name is Peter.
Bot say: Nice to meet you, Peter. Guess what my name is?
User say: Your name is Carl.
Bot say: Wrong guess. My name is Chalmun.

another visit in canteen :-):

User say: Hello, my name is Peter.
Bot say: Nice to meet you, Peter. Guess what my name is?
User say: Chalmun?
Bot say: Yes. That's right.

User say: Do you know who Lando is?
Bot say: I don't know who Lando is.

User Say: Do you know who Joe is?
Bot say: Joe is my neighbor.

another visit in canteen :-):

User say: Hello, my name is Peter.
Bot say: Nice to meet you, Peter. Guess what my name is?
User say: Your name is Chalmun.
Bot say: That's right.

User say: Chalmun
Bot say: Yes. This is me. I am Chalmun.

User say: Chalmun is a nice name.
Bot say: Nice, that you like the name Chalmun.


some other checks with default / emergency answers in the result:
(they are here in german, because I have built a german bot)

User say: Chalmun bla bla bla
Bot Say: ...Mit der Aussage kann ich im Moment nichts anfangen.  

User say: bla bla bla Chalmun
Bot say: ...Mir ist nicht ganz klar, was Sie mir damit sagen möchten.

User say: bla bla Chalmun bla bla
Bot say: ...Das sagt mir gerade nichts.

another dialog with log-file and scoring:

User say: Hello, my name is Peter.
Bot say: Nice to meet you, Peter. Guess what my name is?
User say: Chalmun?
Bot say: Yes. That's right.

log-file:
```
...
19-08-2018 08:30:48.14078[4][4] - Elapsed time: 0.010 milliseconds
[find_aiml.php][score_matches][269]
Scoring the matches. 
-----------------------
19-08-2018 08:30:48.14120[4][4] - Elapsed time: 0.421 milliseconds
[find_aiml.php][score_matches][562]
Sorted array $allrows:
Array
(
    [0] => Array
        (
            [id] => 4264
            [bot_id] => 1
            [pattern] => <BOT NAME="NAME"/>
            [thatpattern] => * GUESS WHAT MY NAME IS
            [topic] => 
            [filename] => test_2.aiml
            [template] => Yes. That's right.
            [score] => 268
            [current_that] => nice to meet you peter guess what my name is
            [aiml_id] => 4264
            [track_score] => current bot (250 points), thatpattern match with star (2 points), uncommon word match: <bot (8 points), uncom
mon word match: name="name"/> (8 points)
        )

    [1] => Array
        (
            [id] => 4262
            [bot_id] => 1
            [pattern] => <BOT NAME="NAME"/>
            [thatpattern] => 
            [topic] => 
            [filename] => test_2.aiml
            [template] => Yes. This is me. I am <bot name="name"/>.
            [score] => 267
            [current_that] => nice to meet you peter guess what my name is
            [aiml_id] => 4262
            [track_score] => current bot (250 points), no thatpattern to match (1 point), uncommon word match: <bot (8 points), uncommon w
ord match: name="name"/> (8 points)
        )

    [2] => Array
        (
            [id] => 2603
            [bot_id] => 1
            [pattern] => *
            [thatpattern] => 
            [topic] => 
            [filename] => Fragen_keine_Ahnung_V1.aiml
            [template] => <srai>RANDOM PICKUP LINE</srai>
            [score] => 252
            [current_that] => nice to meet you peter guess what my name is
            [aiml_id] => 2603
            [track_score] => current bot (250 points), no thatpattern to match (1 point), pattern star match (1 points)
        )

    [3] => Array
        (
            [id] => 2601
            [bot_id] => 1
            [pattern] => RANDOM PICKUP LINE
            [thatpattern] => 
            [topic] => 
            [filename] => Fragen_keine_Ahnung_V1.aiml
            [template] => <random>
   <li>...Mit der Aussage kann ich im Moment nichts anfangen.</li>
   <li>...Das sagt mir gerade nichts.</li>
   <li>...Mir ist nicht ganz klar, was Sie mir damit sagen möchten...</li>
</random>
            [score] => 251
            [current_that] => nice to meet you peter guess what my name is
            [aiml_id] => 2601
            [track_score] => current bot (250 points), no thatpattern to match (1 point)
        )

)
 
-----------------------
19-08-2018 08:30:48.14123[2][4] - Elapsed time: 0.027 milliseconds
[find_aiml.php][get_highest_scoring_row][656]
Final candidates:

Array
(
    [0] => Array
        (
            [id] => 4264
            [bot_id] => 1
            [pattern] => <BOT NAME="NAME"/>
            [thatpattern] => * GUESS WHAT MY NAME IS
            [topic] => 
            [filename] => test_2.aiml
            [template] => Yes. That's right.
            [score] => 268
            [current_that] => nice to meet you peter guess what my name is
            [aiml_id] => 4264
            [track_score] => current bot (250 points), thatpattern match with star (2 points), uncommon word match: <bot (8 points), uncommon word match: name="name"/> (8 points)
        )

)

 
-----------------------
19-08-2018 08:30:48.14149[4][4] - Elapsed time: 0.261 milliseconds
[find_aiml.php][get_highest_scoring_row][711]
Best Response: Array
(
    [id] => 4264
    [bot_id] => 1
    [pattern] => <BOT NAME="NAME"/>
    [thatpattern] => * GUESS WHAT MY NAME IS
    [topic] => 
    [filename] => test_2.aiml
    [template] => Yes. That's right.
    [score] => 268
    [current_that] => nice to meet you peter guess what my name is
    [aiml_id] => 4264
    [track_score] => current bot (250 points), thatpattern match with star (2 points), uncommon word match: <bot (8 points), uncommon word match: name="name"/> (8 points)
)
...
```
Best regards

**Comment by OverlanAI** [comment]
Looks fine, The end use is more about when NPC1 is talking about NPC2 using NPC2's bot name.
Like NPC1 (sue)  "I want a dress like sally has."  NPC2 (sally) can say "Do you like my blue one or the red one, sue ?" ........
And on it goes. 
Why we can do that sort of thing without using a special function, its not as simple as some may think.
Now in that example my program is talking to its self but using different sections of the AIML (not multi-bot mode yet). Enters 8 more NPCs and now they are all talking about whatever and 1 or 2 humans are also talking (works ok btw) but it starts to get very complex keeping track of who said what and who heard what and about what.
So with a few special functions I  can simplify this messy problem quite a bit because the number of who is talking is never a known factor, I might have just 2 or there could be 20 at any given time.
In this test case we are using bot_name (fine have to start with something).
This does not quite go far enough so i would add in one or two more MISC  entrys into the bot profile like
MISC1="Mr." NAME= "Bill". In this case i can combine the 2 in one pattern search (looking for) "Mr. Bill"  not just "Bill". I could name my bot Mr, Bill but think about about that a bit.
Bill and Mr. Bill are still the same thing but a user may or may not add the "Mr." to it. or mess with it like Ms. Bill. 
This gives the AIML a bit of logic where "Bill"=True "Mr. Bill"=True "Ms. Bill" = False and the inverse of 1 like "Bill Mr."=False So with just 2 i can give it 4 logical replies.
If i change the Bot profile values I do not need to edit the aiml or change the logic. Like now my bots name is  SUE .. well you get the idea. For me i will try and make it 3 new values in the profile MISC1-3
With just one special value we don't get too far if you see what i mean.
AIML can make great use of the differences now because they can be split or combined  to form a better reply (logic)  and always can be found in a pattern search.
TNX (testing continues).

**Comment by OverlanAI** [comment]
The ultimate with this for me is to have the special feature let me use any of the bot profile values (why not all of them not just NAME).
Then set a score value to the special feature. BOT NAME="whatever we want" like "BOT NAME=" will give it a score of 10 for example and we ignore the lookup name we are pulling in.
This way we we don't need to hard-code in php a name value since it can be any of them we already set in the profile.
Do we need them all? No but we also don't know what one we might need ether (depends).
I feel if that works now we got a real something that program-O can do well and not break anything in the process  :)
But before we go that far we need to know what risks we have as far as Mysql and php goes, and some bad actor trying to give us a bad day.
TNX

**Comment by OverlanAI** [comment]
So far i have tried adding in other bot=name"XXX" and i faild to make it work. It only found the last XXX i told it to use and ignored the the others even though each on there own did work but not combined.
So i did not get very far in it looking for more than one XXX value.
I do see what we get the bot profile names from the array so that helps at least.
But overall it looks like if we want to deal with tags in the pattern then it may be better to make it a sep. function so that it can handle the XXX we might want and easier to manage.
The XXX is a var. on its own that is in the bot name="xxx" tag. (the idea is so we are not hard coding XXX.
XXX becomes the value of what we set in the AIML pattern and not in core that way.

Testing what you have now (trying to cause problems) like what if the value is null or there is no XXX by a given name had no effect so seems to already be handled anyway. I also tried to cause a recursion condition and was not able to do it. (all good things).

Overall in the form you have it seems quite solid but not as flexible as i wish it could be.
So at least you know about my thinking about it.
Today i am going to have a look at how we can give the tag some kind of score and also because I already was working on that but for other reasons.
My idea for this is to give the tag the same score value as STAR so that it is no more or less powerful then see how that turns out overall.
I feel the use of a TAG should have some weight in the same way a STAR or UNDERSCORE var. would.
Anyway i see no problems in allowing the feature considering using it makes an assumption that the AIML used is a special case where we need to simplify things. And in normal use "AIML bots" changes nothing.
So i will get back to the SCORING for now and let others mull it all this over a few days.

TNX(  Good job man )

**Comment by OverlanAI** [comment]
Okay adding in a score for pattern_bot_name is not a big deal as it turns out.  Can just pull down the same that is used for star or underscore and give it something like $pattern_bot_name with a value of 1.
I am treating the whole bot =name the same as if it were a just a star and the main program doesn't 
 seem to care anyway.
For now i will just score it as a value of 1 but may need adjusted higher if we want to take a bit higher so it will tend to override but i would not try higher than a value of 45 and may not be needed anyway.
Main point is currently i had no problem adding it so the special feature has a bit more weight to it if used.
Tinker around a bit and see what value you come up with.
TNX (totally doable)

**Comment by Josef-Meier** [comment]
Hello
Actually, this should be posted more in a forum as in Github.
My general thoughts on this issue:
Actually, there should be a reasonable possibility that the bot variables retrieved with
 ```<bot name = "xyz" />``` can also be processed in a condition tag. How to do that with user variables (predicate variables).

before few definitions from AIML:
```<bot />```:
The bot element is used to recall custom bot properties defined in the .properties file. These variables are accessible to all users of the bot.

```<set>```:
The set element is used to set a predicate variable. Predicates are not hardcoded like properties, and can be initialized during conversation. This means that input from the user can be echoed in the value of a predicate

Concrete AIML example for user variables (predicate variables):
```
...
<category>
    <pattern> HOW ARE YOU FEELING TODAY </pattern>     
    <template>
        <think><set name = "state"> happy</set></think>
        <condition name = "state" value = "happy">I am happy!</condition>
        <condition name = "state" value = "sad">I am sad!</condition>
      </template>
</category>
...
```
Beispiel für Bot-Variablen (bot properties) :
For this I do not have a valid AIML example yet,
but the following processing logic:
```
User say: I guess you're * years old

Now processing of the user input should be realized as follows:

IF bot="gender" == "female" {
   Bot say: "You do not ask a lady about age."
}

IF bot="gender" == "male" OR "undefined" {
   IF <star/> == bot="age" {Bot say: "Good guess. I am bot="age" years old."}
   IF <star/> !=  bot="age" {Bot say: "You are wrong. I am bot="age" years old."}
}
```
Fictional AIML 3.0 :-) :
```
...
<category>
   <pattern>I GUESS YOU ARE  * YEARS OLD</pattern>
   <condition <bot="gender"/>EQUAL"female"/>You do not ask a lady about age.</condition>
   <condition <bot="gender"/>EQUAL"male"/>
       <condition <bot="age"/>EQUAL"<star/>"/>Good guess. I am bot="age" years old.</condition>
       <condition <bot="age"/>UNEQUAL"<star/>"/>You are wrong. I am bot="age" years old.</condition>
   </condition>
   ...
<category>
...
```
My Program-O version 2.6.10 is for AIML 1.0, as far as I understand it.
As a user, I feel obliged to at least point to issues and try to substantiate these with concrete examples.
But I see not so many chances that features that do not conform to the AIML 1.0 or 1.01 standard will be realized. A new Program-O version for AIML 2.0 has been announced. But that should be more complex as as the current software release.
Until then we can enjoy the current version.
Everything else is in the hands of the developers. And I'm very confident about that.
Best regards

**Comment by OverlanAI** [comment]
`PHP ERROR [Warning] -preg_match(): Unknown modifier '>' in H:\wamp64\www\ProgramO\chatbot\core\aiml\find_aiml.php on Line 495`

Each time the special feature triggers i get the above in the error log.
tnx

**Comment by OverlanAI** [comment]
Ok i think i found away so that Xname in bot name = Xname  can be whatever value we put in the pattern.
"Xname" is whatever we define in the Pattern AIML  like name or age ect..

1st.  Search for the bot name ="Xname" in the pattern
2nd. Extract just Xname from that tag(bot name="xname"). (this is what we want the value of).
3rd. Look up Xname in the bot property's array.
4th. Handle case of not found (ignore). (Means i gave it bad Xname somehow.. does not exist.).
5th. Then replace it as you do now.
Result should be that we can make use of any existing Xname we want in the AIML pattern (bot name=).

What i have not considered yet how to handle export and import yet.
TNX

**Comment by OverlanAI** [comment]
If you look in the find_aiml.php starting at around line 71, You will see that the pattern "<bot" was meant to work but to me it looks incomplete.
So thats a bit of good news in that it was interned to become a wild-card but never gets processed like that further down where it would then lead to it getting a score as a type of wild-card like * or _ is.
A score would also flag it as something of interest and so would temp. hold the id aside as it scores the rest of the input.
Thats why to me it looked like it was intended but never was finished.
My idea of allowing ANY bot name=value can be done but this type of AI is a PUSH type in that the user starts the conversation first. So it would need to search pattern for all <bot ? each round in order to set the bot name value. The result would be an impact on performance.
Anyway with just one known tag as you have the rest can be filled in to the scoring since there is just a fixed value for it to consider if the rest of that is filled in deeper down.

In short it looks to me like they meant for it to work and was never finished.
Anyway i will leave it as that why i Finnish mapping the current scoring system.
TNX (Not everything is bad news).

**Comment by Josef-Meier** [comment]
I'm sorry.
I discovered another problem.
The test cases before that work, but the following test case with star fails.
```
<?xml version="1.0" encoding="UTF-8"?>
<aiml version="1.0">
  <category>
    <pattern><bot name="name"/> IS *</pattern>
    <template>I should be <star/>.</template>
  </category>
</aiml>
```
Example dialog:
Bot name is "Chalmun".
  User say: Chalmun is smart.
  Bot say: I should be.
I expect.
  Bot say: I should be smart.

Get an error like this:
```
[find_aiml.php][Called outside of function][507]
PHP ERROR [Warning] -preg_match(): Unknown modifier '>' in /srv/www/htdocs/nossl/opensim_lsl/ProgramO/chatbot/core/aiml/find_aiml.php on Line 507
```
@OverlanAI: Similar to your error message on line 495.
Let's see if the development department intends to implement the bot tag in pattern.
But I'll be looking for what it could be in my test environment.
Best regards

**Comment by Josef-Meier** [comment]
Because of the test case with "*" which went awry, a little change in the function "unset_all_bad_pattern_matches ()" is necessary.
I have therefore updated my first comment.
Happy testing :smiley:
Best regards

**Comment by OverlanAI** [comment]
Be gone with the evil > (when it hits the bot tag) line 499.
`elseif ($regEx != $category_pattern && (($category_pattern != '*') || ($category_pattern != '_'))) // && preg_match("/$regEx/i", $pattern) != 0) //OLAI change`

try that and see.
The problem starts around line 70 there it check to see if there is a bot tag to include as a wild-card with the ending of ">". preg_match doesn't like that.
PHP suggest that strpos should be used if all we want (in the part i cut off) is a bool.

Since for now the Bot tag is the only tag we have to deal with right now, i just am not going to worry about the Bool and see how it goes.
Anyway with the change there is no more error and seems quite happy sofar.

Also on line 71 i changed it to be
`$startPos = stripos($aiml_pattern, '<bot name');`
so that
`$botTag = substr($aiml_pattern, $startPos, $len);`
Has the whole (correct) Bot tag string we want.
With more fussing i could then just get the Xname from $botTag where more cheap pet tricks could be done.
Anyway you are better at php than i am so im just sharing what i found out.

TNX (I been wrong before and i will be wrong again!)

**Comment by Josef-Meier** [comment]
@OverlanAI 
Hello OverlanAI,
I'm also not such a good php coder. :smiley:
Have also tried your change.
Your change at your line 499
``` elseif ($regEx != $category_pattern && (($category_pattern != '*') || ($category_pattern != '_'))) // && preg_match("/$regEx/i", $pattern) != 0) //OLAI change ```
works when it comes to suppressing the error message.
This is not a problem at the moment, because it checks for a condition if it is matched then executing any code, because empty brackets follow { }.
Here maybe something was kept open by the developer for the future.
But still I had the problem, described in my last test case, that the "*" is not replaced.
That's the reason why I changed something in the function "unset_all_bad_pattern_matches ()".
Here in "pattern" the bot tag ``` <bot name ="XXXX"/>``` is replaced by the actual value from the "Bot Personality". The developer has already done that with the variable $aiml_pattern.
In addition, I have inserted the value in the array $relevantRows under [pattern] with following line.
```  $subrow['pattern'] = $aiml_pattern; ```
So from now on the software uses the actual value from the "Bot Personality" in pattern instead of ``` <bot name ="XXXX"/>``` .
See also my updated first post under **"One"**. Look in source code at the comments with "//// Meier":
After the change also works my last published test case with star.
```
<?xml version="1.0" encoding="UTF-8"?>
<aiml version="1.0">
  <category>
    <pattern><bot name="name"/> IS *</pattern>
    <template>I should be <star/>?</template>
  </category>
</aiml>
```
Example dialog:
Bot name is "Bughunter2".

User say: **Bughunter2 is smart.**
Bot say: **I should be smart?**
before Bot say: **I should be?**

He not process the star.
Think I'll finish the topic now and wait for a reaction from the developers. I do not know if the issue is considered.

Best Regards

**Comment by OverlanAI** [comment]
Ahh ok i i did not see your change, i will do the same here why i wait for the PHP people to tell me how to solve the > issue and so far there suggesting that IF you have to deal with it the function can also use ansi code value to get around the problem IE we swap out the > for the ansi code then swap it back once it is done with the regex preg_match.
However they say that using the srtpos is better if all we want is the Bool of the match and that looks to be the case here.
Tomorrow i will change to strpos so that i can add in the score bits that are missing and i think thats where dave might have gotten suck given the {} do nothing.
Also hippo found a few places where global should be static types in order to prevent possible value corruption but thats minor.

Anyway I currently have 4 different bot tags now, not just one and they all work so today im just going to play with the new toys and set the code aside.
I added in 4 Misc bot personality entries and gave them each there own bot tag so they can be anything i want them to be.
Now i can play around and see how useful all this is :-)

**Comment by OverlanAI** [comment]
Haha just noticed "Bughunter2", thought that was cool.
I think ill make an NPC called "Bughunter" just for the heck of it lol.
TNX

**Comment by OverlanAI** [comment]
Adding `$subrow['pattern'] = $aiml_pattern;` is not good for me with more than one bot tag type so i backed that out for now.
It will only use the first one it finds and causes it to loop a bit where `$botTag=` becomes duplicated.
Anyway not having a combo of * and a bot name = in the same AIML pattern is not that big of a deal to me.
I would rather have many bot name = when i might want to use the function and that is i rare cases anyway.
So on that point we tend to split in our ideas a bit, but its all good stuff.
tnx

**Comment by OverlanAI** [comment]
Now i know others are thinking why the heck would you want to do this anyway? Why not just
`* BILL AND BOB` and be done with it.
The main reason is a difference in having to edit an aiml set vrs just a fast change in the personality.
If i change BILL to be SALLY in the current way means i have to change each pattern in the AIML vrs this way where i can simply do it in the bots profile.
When we get into a complex bot this can be a big time saver.
Also when you are dealing with multi-bots this is quite handy.
So all in all it saves time when things get complex.
TNX

**Comment by OverlanAI** [comment]
https://stackoverflow.com/questions/8656089/
If your getting nesting level errors.
I think this may explain a lot when it comes to WAMP anyway.

**Comment by OverlanAI** [comment]
I have managed to get this to work quite well now but I am questioning if this is the best place to have a TAG that is also a Variable.
We do have the ability to make our own custom tags now but There is not a lot of information about it.
If that is the best place for things like this, would it work in the pattern to ?
Anyway i did replace the occurrences of preg_match using the PHP suggested change to srtpos where we just need a bool check result and does not care about > in the array.
It is faster than using "preg_match" that is used in a few places to do the same thing.
The logic is the same and wont throw a fit about a > in it.
You will see that it is also in the that-pattern and it also sees the > there, besides in the current "do nothing".
Anyway after i did that i had no more problems with >.
Also adding `$subrow['pattern'] = $aiml_pattern;` is not a good idea but i it does work.
The problem with doing it there is "recursion" and cause it to be slow on each pass. Sadly i did not find a better way so currently i just keep in mind that any * value will not transfer to the template for now.
Mostly Beware of the risks of "recursion in php".
TNX
